java-checks/src/main/java/org/sonar/java/checks/AbstractAccessibilityChangeChecker.java

```diff
@@ -0,0 +1,148 @@
+    MethodMatchers.create().ofTypes("java.lang.Class")
+      .names("getFields", "getDeclaredFields")
+      .addWithoutParametersMatcher()
+      .build()
+  );
+
+  @Override
+  protected MethodMatchers getMethodInvocationMatchers() {
+    return MethodMatchers.or(
+      SET_MATCHERS,
+      SET_ACCESSIBLE_MATCHER
+    );
+  }
+
+  protected static boolean setsToPubliclyAccessible(MethodInvocationTree mit) {
+    Arguments arguments = mit.arguments();
+    ExpressionTree arg = arguments.get(0);
+    if (arguments.size() > 1) {
+      arg = arguments.get(1);
+    }
+    return Boolean.TRUE.equals(ExpressionsHelper.getConstantValueAsBoolean(arg).value());
+  }
+
+  protected static boolean isModifyingFieldFromRecord(MethodInvocationTree mit) {
+    ExpressionTree fieldModificationExpression = mit.methodSelect();
+    if (!fieldModificationExpression.is(Tree.Kind.MEMBER_SELECT)) {
+      return false;
+    }
+    MemberSelectExpressionTree memberSelect = (MemberSelectExpressionTree) fieldModificationExpression;
+    ExpressionTree expression = memberSelect.expression();
+    if (expression.is(Tree.Kind.ARRAY_ACCESS_EXPRESSION)) {
+      ArrayAccessExpressionTree arrayAccess = (ArrayAccessExpressionTree) expression;
+      expression = arrayAccess.expression();
+    }
+    MethodInvocationTree fieldInitializer = null;
+    if (expression.is(Tree.Kind.METHOD_INVOCATION)) {
+      fieldInitializer = (MethodInvocationTree) expression;
+    } else if (expression.is(Tree.Kind.IDENTIFIER)) {
+      IdentifierTree identifier = (IdentifierTree) expression;
+      Optional<MethodInvocationTree> fieldGettingInvocation = getInitializingMethodInvocation(identifier);
+      if (!fieldGettingInvocation.isPresent()) {
+        return false;
+      }
+      fieldInitializer = fieldGettingInvocation.get();
+    } else {
+      return false;
+    }
+    if (!FIELD_FETCHING_METHODS.matches(fieldInitializer)) {
+      return false;
+    }
+    ExpressionTree initializerExpression = fieldInitializer.methodSelect();
+    Optional<IdentifierTree> classIdentifier = getClassIdentifier((MemberSelectExpressionTree) initializerExpression);
+    if (!classIdentifier.isPresent()) {
+      return false;
+    }
+    return classIdentifier.get().symbol().type().isSubtypeOf("java.lang.Record");
+  }
+
+  private static Optional<MethodInvocationTree> getInitializingMethodInvocation(IdentifierTree identifier) {
+    Tree declaration = identifier.symbol().declaration();
+    if (declaration == null || !declaration.is(Tree.Kind.VARIABLE)) {
+      return Optional.empty();
+    }
+    VariableTree variable = (VariableTree) declaration;
+    ExpressionTree initializer = variable.initializer();
+    if (initializer == null || !initializer.is(Tree.Kind.METHOD_INVOCATION)) {
+      return Optional.empty();
+    }
+    return Optional.of((MethodInvocationTree) initializer);
+  }
+
+  private static Optional<IdentifierTree> getClassIdentifier(MemberSelectExpressionTree memberSelect) {
```

Comment:
I kind of want to replace this entire method with simply checking with simply getting the symbol type of the expression that `getField` is called on and asking it whether its type argument inherits `Record`. This wouldn't work in the case where the `Class` object gets passed around as `Class<? extends Object>`, but the current approach doesn't work in that case either (indeed the current way doesn't work in any case where the `Class` object is being passed around).

Sadly, type symbols don't currently know their type arguments, so it's not that simple. The ECJ API does have that information though, so we might be able to add it to `TypeSymbol` without too much trouble.